File: forms/sports-medicine-assessment/full-stack-with-loco-tera-htmx-alpine/tests/engine/ppe_grader_test.rs
use sports_medicine_assessment_tera_crate::engine::ppe_grader::{
    aggregate_clearance, grade, run_rules,
};
use sports_medicine_assessment_tera_crate::engine::ppe_rules::all_rules;
use sports_medicine_assessment_tera_crate::engine::types::*;

/// Build a "healthy male athlete, no concerns" baseline that should result in
/// a Cleared clearance with no rules fired.
fn cleared_case() -> AssessmentData {
    AssessmentData {
        demographics: Demographics {
            first_name: "Alex".to_string(),
            last_name: "Doe".to_string(),
            date_of_birth: "2002-05-12".to_string(),
            sex: "male".to_string(),
            weight: Some(75.0),
            height: Some(180.0),
            bmi: Some(23.1),
            emergency_contact_name: "Pat Doe".to_string(),
            emergency_contact_phone: "555-1234".to_string(),
        },
        sport_position_details: SportPositionDetails {
            primary_sport: "Soccer".to_string(),
            primary_position: "Midfielder".to_string(),
            contact_level: "moderate".to_string(),
            secondary_sports: String::new(),
            competitive_level: "club".to_string(),
            hours_per_week: Some(10.0),
            previous_clearance_issue: "no".to_string(),
            previous_clearance_details: String::new(),
        },
        medical_history: MedicalHistory {
            chronic_illness: "no".to_string(),
            current_medications: "no".to_string(),
            allergies_known: "no".to_string(),
            prior_surgery: "no".to_string(),
            hospitalised_last_year: "no".to_string(),
            asthma_or_exercise_induced_bronchospasm: "no".to_string(),
            diabetes: "no".to_string(),
            sickle_cell_trait_or_disease: "no".to_string(),
            heat_illness_history: "no".to_string(),
            eating_disorder_history: "no".to_string(),
            ..MedicalHistory::default()
        },
        family_history: FamilyHistory {
            sudden_cardiac_death_under_50: "no".to_string(),
            hypertrophic_cardiomyopathy: "no".to_string(),
            marfan_syndrome: "no".to_string(),
            long_qt_syndrome: "no".to_string(),
            arrhythmia_or_pacemaker: "no".to_string(),
            unexplained_seizure_or_fainting: "no".to_string(),
            ..FamilyHistory::default()
        },
        cardiovascular_screening: CardiovascularScreening {
            chest_pain_with_exertion: "no".to_string(),
            unexplained_syncope: "no".to_string(),
            excessive_breathlessness: "no".to_string(),
            palpitations_or_irregular_beat: "no".to_string(),
            high_blood_pressure_diagnosis: "no".to_string(),
            heart_murmur_detected: "no".to_string(),
            restricted_activity_for_heart: "no".to_string(),
            resting_systolic: Some(118),
            resting_diastolic: Some(76),
            resting_heart_rate: Some(62),
        },
        musculoskeletal_screening: MusculoskeletalScreening {
            uncorrected_major_injury: "no".to_string(),
            joint_instability: "no".to_string(),
            ongoing_pain_or_swelling: "no".to_string(),
            chronic_joint_disease: "no".to_string(),
            use_brace_or_assistive_device: "no".to_string(),
            full_range_of_motion: "yes".to_string(),
            normal_strength_bilateral: "yes".to_string(),
            ..MusculoskeletalScreening::default()
        },
        neurological_concussion_baseline: NeurologicalConcussionBaseline {
            total_concussions: Some(0),
            concussion_last_six_months: "no".to_string(),
            ongoing_post_concussive_symptoms: "no".to_string(),
            history_of_seizures: "no".to_string(),
            stinger: "no".to_string(),
            history_of_head_or_neck_surgery: "no".to_string(),
            baseline_headaches_or_migraine: "no".to_string(),
            ..NeurologicalConcussionBaseline::default()
        },
        vision_skin: VisionSkin {
            corrective_lenses_worn: "no".to_string(),
            monocular_athlete: "no".to_string(),
            protective_eyewear_available: "yes".to_string(),
            active_skin_infection: "no".to_string(),
            herpes_gladiatorum: "no".to_string(),
            impetigo_or_mrsa: "no".to_string(),
            open_wounds_or_lesions: "no".to_string(),
            ..VisionSkin::default()
        },
        clearance_decision: ClearanceDecision {
            preferred_clearance: "cleared".to_string(),
            clinician_name: "Dr Casey".to_string(),
            clinician_signature_date: "2026-06-01".to_string(),
            ..ClearanceDecision::default()
        },
        ..AssessmentData::default()
    }
}

#[test]
fn empty_assessment_yields_cleared_when_no_rules_fire() {
    // With all enums empty, no rule fires; aggregate yields "cleared".
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.clearance, "cleared");
    assert!(result.fired_rules.is_empty());
}

#[test]
fn baseline_healthy_athlete_is_cleared() {
    let data = cleared_case();
    let result = grade(&data);
    assert_eq!(result.clearance, "cleared");
    assert!(result.fired_rules.is_empty(), "{:?}", result.fired_rules);
}

#[test]
fn chest_pain_with_exertion_is_not_cleared() {
    let mut data = cleared_case();
    data.cardiovascular_screening.chest_pain_with_exertion = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.clearance, "not-cleared");
    assert!(result.fired_rules.iter().any(|r| r.id == "PPE-030"));
}

#[test]
fn family_scd_under_50_is_not_cleared() {
    let mut data = cleared_case();
    data.family_history.sudden_cardiac_death_under_50 = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.clearance, "not-cleared");
    assert!(result.fired_rules.iter().any(|r| r.id == "PPE-010"));
}

#[test]
fn ongoing_post_concussive_symptoms_is_not_cleared() {
    let mut data = cleared_case();
    data.neurological_concussion_baseline.ongoing_post_concussive_symptoms = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.clearance, "not-cleared");
    assert!(result.fired_rules.iter().any(|r| r.id == "PPE-052"));
}

#[test]
fn chronic_illness_alone_is_conditional() {
    let mut data = cleared_case();
    data.medical_history.chronic_illness = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.clearance, "conditional");
    assert!(result.fired_rules.iter().any(|r| r.id == "PPE-001"));
}

#[test]
fn marfan_alone_is_pending() {
    let mut data = cleared_case();
    data.family_history.marfan_syndrome = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.clearance, "pending");
    assert!(result.fired_rules.iter().any(|r| r.id == "PPE-012"));
}

#[test]
fn reds_triad_in_female_athlete_is_not_cleared() {
    let mut data = cleared_case();
    data.demographics.sex = "female".to_string();
    data.menstrual_history_reds = MenstrualHistoryReds {
        applicable: true,
        amenorrhoea_six_months: "yes".to_string(),
        restrictive_eating_pattern: "yes".to_string(),
        stress_fracture_history: "yes".to_string(),
        ..MenstrualHistoryReds::default()
    };
    let result = grade(&data);
    assert_eq!(result.clearance, "not-cleared");
    assert!(result.fired_rules.iter().any(|r| r.id == "PPE-023"));
}

#[test]
fn high_resting_bp_fires_ppe_036() {
    let mut data = cleared_case();
    data.cardiovascular_screening.resting_systolic = Some(150);
    data.cardiovascular_screening.resting_diastolic = Some(95);
    let fired = run_rules(&data);
    assert!(fired.iter().any(|r| r.id == "PPE-036"));
}

#[test]
fn three_lifetime_concussions_fires_ppe_051() {
    let mut data = cleared_case();
    data.neurological_concussion_baseline.total_concussions = Some(3);
    let fired = run_rules(&data);
    assert!(fired.iter().any(|r| r.id == "PPE-051"));
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

#[test]
fn aggregate_clearance_picks_highest_grade() {
    let rules = vec![
        FiredRule {
            id: "A".to_string(),
            category: "x".to_string(),
            description: "".to_string(),
            grade: 2,
        },
        FiredRule {
            id: "B".to_string(),
            category: "x".to_string(),
            description: "".to_string(),
            grade: 3,
        },
    ];
    assert_eq!(aggregate_clearance(&rules), "pending");
}

File: forms/sports-medicine-assessment/front-end-with-svelte/src/lib/engine/ppe-grader.test.ts
import { describe, it, expect } from 'vitest';
import { gradePPE, aggregateClearance, countAnswered } from './ppe-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { ppeRules } from './ppe-rules';
import type { AssessmentData } from './types';

/**
 * A blank assessment with all fields at their unanswered defaults. Mirrors the
 * store's `createDefaultAssessment`, but inlined so this engine test stays pure
 * (no SvelteKit `$app/*` imports).
 */
function createDefaultAssessment(): AssessmentData {
	return {
		demographics: {
			firstName: '',
			lastName: '',
			dateOfBirth: '',
			sex: '',
			weight: null,
			height: null,
			bmi: null,
			emergencyContactName: '',
			emergencyContactPhone: ''
		},
		sportPositionDetails: {
			primarySport: '',
			primaryPosition: '',
			contactLevel: '',
			secondarySports: '',
			competitiveLevel: '',
			hoursPerWeek: null,
			previousClearanceIssue: '',
			previousClearanceDetails: ''
		},
		medicalHistory: {
			chronicIllness: '',
			chronicIllnessDetails: '',
			currentMedications: '',
			currentMedicationDetails: '',
			allergiesKnown: '',
			allergyDetails: '',
			priorSurgery: '',
			priorSurgeryDetails: '',
			hospitalisedLastYear: '',
			asthmaOrExerciseInducedBronchospasm: '',
			diabetes: '',
			sickleCellTraitOrDisease: '',
			heatIllnessHistory: '',
			eatingDisorderHistory: ''
		},
		familyHistory: {
			suddenCardiacDeathUnder50: '',
			suddenCardiacDeathRelation: '',
			hypertrophicCardiomyopathy: '',
			marfanSyndrome: '',
			longQTSyndrome: '',
			arrhythmiaOrPacemaker: '',
			unexplainedSeizureOrFainting: ''
		},
		menstrualHistoryREDS: {
			applicable: false,
			ageAtMenarche: null,
			regularPeriods: '',
			amenorrhoeaSixMonths: '',
			cyclesLast12Months: null,
			restrictiveEatingPattern: '',
			stressFractureHistory: '',
			lowEnergyAvailabilityConcern: ''
		},
		cardiovascularScreening: {
			chestPainWithExertion: '',
			unexplainedSyncope: '',
			excessiveBreathlessness: '',
			palpitationsOrIrregularBeat: '',
			highBloodPressureDiagnosis: '',
			heartMurmurDetected: '',
			restrictedActivityForHeart: '',
			restingSystolic: null,
			restingDiastolic: null,
			restingHeartRate: null
		},
		musculoskeletalScreening: {
			uncorrectedMajorInjury: '',
			majorInjuryDetails: '',
			jointInstability: '',
			jointInstabilityDetails: '',
			ongoingPainOrSwelling: '',
			chronicJointDisease: '',
			useBraceOrAssistiveDevice: '',
			fullRangeOfMotion: '',
			normalStrengthBilateral: ''
		},
		neurologicalConcussionBaseline: {
			totalConcussions: null,
			concussionLastSixMonths: '',
			mostRecentConcussionDate: '',
			ongoingPostConcussiveSymptoms: '',
			historyOfSeizures: '',
			stinger: '',
			historyOfHeadOrNeckSurgery: '',
			baselineHeadachesOrMigraine: ''
		},
		visionSkin: {
			correctiveLensesWorn: '',
			monocularAthlete: '',
			protectiveEyewearAvailable: '',
			activeSkinInfection: '',
			activeSkinInfectionDetails: '',
			herpesGladiatorum: '',
			impetigoOrMRSA: '',
			openWoundsOrLesions: ''
		},
		clearanceDecision: {
			preferredClearance: '',
			clearanceConditions: '',
			followUpRequired: '',
			clinicianName: '',
			clinicianSignatureDate: '',
			additionalNotes: ''
		}
	};
}

/** A fully-answered, low-risk athlete with no concerns. */
function healthyAthlete(): AssessmentData {
	const d = createDefaultAssessment();
	d.demographics = {
		...d.demographics,
		firstName: 'Alex',
		lastName: 'Carter',
		dateOfBirth: '2002-03-14',
		sex: 'male',
		weight: 75,
		height: 180,
		bmi: 23.1
	};
	d.sportPositionDetails = {
		...d.sportPositionDetails,
		primarySport: 'Athletics',
		contactLevel: 'low',
		competitiveLevel: 'club',
		hoursPerWeek: 8,
		previousClearanceIssue: 'no'
	};
	d.medicalHistory = {
		...d.medicalHistory,
		chronicIllness: 'no',
		currentMedications: 'no',
		allergiesKnown: 'no',
		priorSurgery: 'no',
		hospitalisedLastYear: 'no',
		asthmaOrExerciseInducedBronchospasm: 'no',
		diabetes: 'no',
		sickleCellTraitOrDisease: 'no',
		heatIllnessHistory: 'no',
		eatingDisorderHistory: 'no'
	};
	d.familyHistory = {
		...d.familyHistory,
		suddenCardiacDeathUnder50: 'no',
		hypertrophicCardiomyopathy: 'no',
		marfanSyndrome: 'no',
		longQTSyndrome: 'no',
		arrhythmiaOrPacemaker: 'no',
		unexplainedSeizureOrFainting: 'no'
	};
	d.cardiovascularScreening = {
		...d.cardiovascularScreening,
		chestPainWithExertion: 'no',
		unexplainedSyncope: 'no',
		excessiveBreathlessness: 'no',
		palpitationsOrIrregularBeat: 'no',
		highBloodPressureDiagnosis: 'no',
		heartMurmurDetected: 'no',
		restrictedActivityForHeart: 'no'
	};
	d.musculoskeletalScreening = {
		...d.musculoskeletalScreening,
		uncorrectedMajorInjury: 'no',
		jointInstability: 'no',
		ongoingPainOrSwelling: 'no',
		chronicJointDisease: 'no',
		fullRangeOfMotion: 'yes',
		normalStrengthBilateral: 'yes'
	};
	d.neurologicalConcussionBaseline = {
		...d.neurologicalConcussionBaseline,
		totalConcussions: 0,
		concussionLastSixMonths: 'no',
		ongoingPostConcussiveSymptoms: 'no',
		historyOfSeizures: 'no',
		stinger: 'no',
		historyOfHeadOrNeckSurgery: 'no'
	};
	d.visionSkin = {
		...d.visionSkin,
		correctiveLensesWorn: 'no',
		monocularAthlete: 'no',
		activeSkinInfection: 'no',
		herpesGladiatorum: 'no',
		impetigoOrMRSA: 'no',
		openWoundsOrLesions: 'no'
	};
	return d;
}

describe('PPE clearance grader', () => {
	it('clears a healthy athlete with no fired rules', () => {
		const result = gradePPE(healthyAthlete());
		expect(result.clearance).toBe('cleared');
		expect(result.firedRules).toHaveLength(0);
		expect(result.answeredCount).toBeGreaterThan(0);
	});

	it('returns conditional when only grade-2 rules fire', () => {
		const d = healthyAthlete();
		d.medicalHistory.asthmaOrExerciseInducedBronchospasm = 'yes';
		const result = gradePPE(d);
		expect(result.clearance).toBe('conditional');
	});

	it('returns pending when a grade-3 rule fires', () => {
		const d = healthyAthlete();
		d.neurologicalConcussionBaseline.concussionLastSixMonths = 'yes';
		const result = gradePPE(d);
		expect(result.clearance).toBe('pending');
	});

	it('returns not-cleared for an exertional cardiac red flag', () => {
		const d = healthyAthlete();
		d.cardiovascularScreening.chestPainWithExertion = 'yes';
		const result = gradePPE(d);
		expect(result.clearance).toBe('not-cleared');
		expect(result.firedRules.some((r) => r.id === 'PPE-030')).toBe(true);
	});

	it('excludes infectious skin lesions only in contact sport', () => {
		const d = healthyAthlete();
		d.visionSkin.impetigoOrMRSA = 'yes';
		// Low-contact sport: rule must not fire.
		expect(gradePPE(d).firedRules.some((r) => r.id === 'PPE-070')).toBe(false);
		// Contact sport: rule fires and excludes.
		d.sportPositionDetails.contactLevel = 'high';
		const result = gradePPE(d);
		expect(result.firedRules.some((r) => r.id === 'PPE-070')).toBe(true);
		expect(result.clearance).toBe('not-cleared');
	});

	it('aggregates the worst grade across fired rules', () => {
		expect(aggregateClearance([])).toBe('cleared');
		expect(aggregateClearance([{ id: 'x', category: 'c', description: 'd', grade: 2 }])).toBe(
			'conditional'
		);
		expect(
			aggregateClearance([
				{ id: 'x', category: 'c', description: 'd', grade: 2 },
				{ id: 'y', category: 'c', description: 'd', grade: 4 }
			])
		).toBe('not-cleared');
	});

	it('has unique rule ids', () => {
		const ids = ppeRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});

	it('counts RED-S fields only when applicable', () => {
		const base = healthyAthlete();
		const baseline = countAnswered(base);
		const withReds = healthyAthlete();
		withReds.menstrualHistoryREDS.applicable = true;
		withReds.menstrualHistoryREDS.regularPeriods = 'yes';
		expect(countAnswered(withReds)).toBeGreaterThan(baseline);
	});
});

describe('PPE flagged-issue detection', () => {
	it('raises no flags for a healthy athlete', () => {
		expect(detectAdditionalFlags(healthyAthlete())).toHaveLength(0);
	});

	it('flags a recent concussion as high priority', () => {
		const d = healthyAthlete();
		d.neurologicalConcussionBaseline.concussionLastSixMonths = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-NEURO-001' && f.priority === 'high')).toBe(true);
	});

	it('flags the probable RED-S triad when applicable', () => {
		const d = healthyAthlete();
		d.demographics.sex = 'female';
		d.demographics.bmi = 17;
		d.menstrualHistoryREDS.applicable = true;
		d.menstrualHistoryREDS.amenorrhoeaSixMonths = 'yes';
		d.menstrualHistoryREDS.stressFractureHistory = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-REDS-001')).toBe(true);
	});

	it('sorts flags high → medium → low', () => {
		const d = healthyAthlete();
		d.cardiovascularScreening.chestPainWithExertion = 'yes'; // high
		d.medicalHistory.heatIllnessHistory = 'yes'; // medium
		d.sportPositionDetails.hoursPerWeek = 25; // low
		const order = { high: 0, medium: 1, low: 2 };
		const priorities = detectAdditionalFlags(d).map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});
});
